File: simulation/agent/Agent.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.vaticle.typedb.benchmark.simulation.agent;

import com.vaticle.factory.tracing.client.FactoryTracingThreadStatic;
import com.vaticle.typedb.benchmark.common.concept.Region;
import com.vaticle.typedb.benchmark.common.params.Context;
import com.vaticle.typedb.benchmark.common.seed.RandomSource;
import com.vaticle.typedb.benchmark.simulation.driver.Client;
import com.vaticle.typedb.benchmark.simulation.driver.Session;
import com.vaticle.typedb.benchmark.simulation.driver.Transaction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Supplier;

import static com.vaticle.factory.tracing.client.FactoryTracingThreadStatic.contextOnThread;
import static com.vaticle.factory.tracing.client.FactoryTracingThreadStatic.traceOnThread;
import static com.vaticle.typedb.common.collection.Collections.pair;
import static com.vaticle.typedb.common.util.Objects.className;
import static java.util.concurrent.CompletableFuture.runAsync;

/**
 * Agent constructs regional agents of a given class and runs them in parallel, providing them with the appropriate
 * region, a deterministic random and the tracker and session key for tracing and TypeDB transactions.
 *
 * This class must be extended to provide the source of the random items and the methods to obtain the session key and
 * tracker from them.
 *
 * @param <REGION> The type of region used by the agent.
 * @param <TX>     The abstraction of database operations used by the agent.
 */
public abstract class Agent<REGION extends Region, TX extends Transaction> {

    private static final Logger LOG = LoggerFactory.getLogger(Agent.class);

    protected final Context context;
    private final Client<?, TX> client;
    private boolean isTracing = true;

    protected Agent(Client<?, TX> client, Context context) {
        this.client = client;
        this.context = context;
    }

    protected abstract Class<? extends Agent> agentClass();

    protected abstract List<REGION> regions();

    protected abstract List<Report> run(Session<TX> session, REGION region, RandomSource random);

    public Agent<REGION, TX> setTracing(boolean isTracing) {
        this.isTracing = isTracing;
        return this;
    }

    public boolean isTracing() {
        return context.isTracing() && isTracing;
    }

    public Map<String, List<Report>> iterate(RandomSource randomSrc) {
        ConcurrentMap<String, List<Report>> reports = new ConcurrentHashMap<>();
        List<CompletableFuture<Void>> asyncRuns = new ArrayList<>(regions().size());
        // We need to generate pairs of Region and Random deterministically before passing them to a parallel stream
        regions().stream().map(r -> pair(r, randomSrc.nextSource())).forEach(rr -> asyncRuns.add(runAsync(() -> {
            List<Report> report = runAndMayTrace(rr.first(), rr.second());
            if (context.isReporting()) reports.put(rr.first().tracker(), report);
            else assert report.isEmpty();
        }, context.executor())));
        CompletableFuture.allOf(asyncRuns.toArray(new CompletableFuture[0])).join();
        return reports;
    }

    private List<Report> runAndMayTrace(REGION region, RandomSource random) {
        FactoryTracingThreadStatic.ThreadContext tracingCtx = null;
        try {
            if (isTracing()) tracingCtx = contextOnThread(region.tracker(), context.iterationNumber());
            Session<TX> session = client.session(region);
            return mayTrace(() -> run(session, region, random), className(agentClass()));
        } finally {
            if (tracingCtx != null) tracingCtx.close();
        }
    }

    public <T> T mayTrace(Supplier<T> methodToTrace, String trace) {
        if (isTracing()) {
            try (FactoryTracingThreadStatic.ThreadTrace ignored = traceOnThread(trace)) {
                return methodToTrace.get();
            }
        } else {
            return methodToTrace.get();
        }
    }

    public static class Report {

        private final Collection<Object> input;
        private final Collection<Object> output;
        private final int hash;

        public Report(Collection<Object> input, Collection<Object> output) {
            this.input = input;
            this.output = output;
            this.hash = Objects.hash(this.input, this.output);
        }

        public static Report create(Collection<Object> input, Collection<Object> output) {
            return new Report(input, output);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Report report = (Report) o;

            if (!input.equals(report.input)) return false;
            return output.equals(report.output);
        }

        @Override
        public int hashCode() {
            return hash;
        }
    }
}

File: simulation/agent/Agent.kt
/*
 * Copyright (C) 2022 Vaticle
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.vaticle.typedb.benchmark.simulation.agent

import com.vaticle.factory.tracing.client.FactoryTracingThreadStatic
import com.vaticle.typedb.benchmark.common.concept.Region
import com.vaticle.typedb.benchmark.common.params.Context
import com.vaticle.typedb.benchmark.common.seed.RandomSource
import com.vaticle.typedb.benchmark.simulation.driver.Client
import com.vaticle.typedb.benchmark.simulation.driver.Session
import com.vaticle.typedb.benchmark.simulation.driver.Transaction
import com.vaticle.typedb.common.collection.Pair
import com.vaticle.typedb.common.util.Objects.className
import java.util.Objects.hash
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Supplier

/**
 * Agent constructs regional agents of a given class and runs them in parallel, providing them with the appropriate
 * region, a deterministic random and the tracker and session key for tracing and TypeDB transactions.
 *
 * This class must be extended to provide the source of the random items and the methods to obtain the session key and
 * tracker from them.
 *
 * @param <REGION> The type of region used by the agent.
 * @param <TX>     The abstraction of database operations used by the agent.
</TX></REGION> */
abstract class Agent<REGION: Region, TX: Transaction> protected constructor(
    private val client: Client<Session<TX>>, protected val context: Context
) {
    var tracingEnabled = true
    protected abstract val agentClass: Class<out Agent<*, *>>
    protected abstract val regions: List<REGION>
    protected abstract fun run(session: Session<TX>, region: REGION, random: RandomSource): List<Report>

    private fun shouldTrace(): Boolean {
        return context.isTracing && this.tracingEnabled
    }

    fun iterate(randomSrc: RandomSource): Map<String, List<Report>> {
        val reports = ConcurrentHashMap<String, List<Report>>()
        val asyncRuns = mutableListOf<CompletableFuture<Void>>()
        // We need to generate pairs of Region and Random deterministically before passing them to a parallel stream
        regions.map { r: REGION -> Pair(r, randomSrc.nextSource()) }
            .forEach { rr: Pair<REGION, RandomSource> ->
                asyncRuns.add(CompletableFuture.runAsync(
                    {
                        val report = runAndMayTrace(rr.first(), rr.second())
                        if (context.isReporting) reports[rr.first().tracker] = report else assert(report.isEmpty())
                    }, context.executor
                ))
            }
        CompletableFuture.allOf(*asyncRuns.toTypedArray<CompletableFuture<*>>()).join()
        return reports
    }

    private fun runAndMayTrace(region: REGION, random: RandomSource): List<Report> {
        var tracingCtx: FactoryTracingThreadStatic.ThreadContext? = null
        return try {
            if (shouldTrace()) tracingCtx =
                FactoryTracingThreadStatic.contextOnThread(region.tracker, context.iterationNumber)
            val session = client.session(region)
            mayTrace(className(agentClass)) { run(session, region, random) }
        } finally {
            tracingCtx?.close()
        }
    }

    private fun <T> mayTrace(trace: String?, methodToTrace: Supplier<T>): T {
        if (shouldTrace()) {
            FactoryTracingThreadStatic.traceOnThread(trace).use { return methodToTrace.get() }
        } else {
            return methodToTrace.get()
        }
    }

    class Report(private val input: Collection<Any>, private val output: Collection<Any>) {
        private val hash = hash(input, output)

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other == null || javaClass != other.javaClass) return false
            val report = other as Report
            return if (input != report.input) false else output == report.output
        }

        override fun hashCode(): Int {
            return hash
        }
    }
}
